### ISSUE TITLE
Getting `Cannot resolve @babel/runtime/helpers/interopRequireDefault` while initial setup of expo with nx

### ISSUE THREAD

**Issue by sandheep45** [nrwl/nx#19905]
### Current Behavior

Unable to spin up `expo` server  with freshly created nx monorepo using command `pnpx create-nx-workspace --pm pnpm --workspaceType=package-based `

![Screenshot from 2023-10-27 22-18-28](https://github.com/nrwl/nx/assets/105055197/a1c2323f-4bd0-4a66-8a81-d2571eec56c4)

![Screenshot from 2023-10-27 22-20-36](https://github.com/nrwl/nx/assets/105055197/e4e490df-19be-46f2-809f-3e8603d76512)

![Screenshot from 2023-10-27 22-21-56](https://github.com/nrwl/nx/assets/105055197/4f43e3d5-019c-495b-90ff-8fab60757865)

### Expected Behavior

To start `expo` server with `nx start <project>`

### Steps to Reproduce

1. Init `nx` monorepo using `pnpx create-nx-workspace --pm pnpm --workspaceType=package-based ` command
2. Filling up command as shown in the image
![Screenshot from 2023-10-27 22-18-28](https://github.com/nrwl/nx/assets/105055197/e03223fa-dbb4-42a6-941e-3ac5ef880077)
3. Navigate into the directory `cd shopnest`
4. Fix unmet package version using `pnpm nx install mobile --fix`
5. Starting up expo server `pnpm nx start mobile`

### Nx Report

```shell
NX   Report complete - copy this into the issue template

   Node   : 18.18.2
   OS     : linux-x64
   pnpm   : 8.9.2
   
   nx                 : 17.0.2
   @nx/js             : 17.0.2
   @nx/jest           : 17.0.2
   @nx/linter         : 17.0.2
   @nx/eslint         : 17.0.2
   @nx/workspace      : 17.0.2
   @nx/detox          : 17.0.2
   @nx/devkit         : 17.0.2
   @nx/eslint-plugin  : 17.0.2
   @nx/expo           : 17.0.2
   @nx/react          : 17.0.2
   @nrwl/tao          : 17.0.2
   @nx/web            : 17.0.2
   @nx/webpack        : 17.0.2
   typescript         : 5.1.6
```
```


### Failure Logs

```shell
error: Error: Cannot resolve @babel/runtime/helpers/interopRequireDefault
    at /home/sandheep/projects/personal/shopnest/node_modules/.pnpm/@nx+expo@17.0.2_@expo+cli@0.10.14_@swc-node+register@1.6.8_@swc+core@1.3.95_@types+node@16.11_ixupc4dnoyqo63cnmrjaebhgma/node_modules/@nx/expo/plugins/metro-resolver.js:30:15
    at resolveRequest (/home/sandheep/projects/personal/shopnest/node_modules/.pnpm/@expo+cli@0.10.14_expo-modules-autolinking@1.5.1/node_modules/@expo/cli/build/src/start/server/metro/withMetroResolvers.js:50:24)
    at Object.resolve (/home/sandheep/projects/personal/shopnest/node_modules/.pnpm/metro-resolver@0.76.7/node_modules/metro-resolver/src/resolve.js:47:12)
    at ModuleResolver.resolveDependency (/home/sandheep/projects/personal/shopnest/node_modules/.pnpm/metro@0.76.7/node_modules/metro/src/node-haste/DependencyGraph/ModuleResolution.js:79:31)
    at DependencyGraph.resolveDependency (/home/sandheep/projects/personal/shopnest/node_modules/.pnpm/metro@0.76.7/node_modules/metro/src/node-haste/DependencyGraph.js:277:43)
    at Object.resolve (/home/sandheep/projects/personal/shopnest/node_modules/.pnpm/metro@0.76.7/node_modules/metro/src/lib/transformHelpers.js:169:21)
    at Graph._resolveDependencies (/home/sandheep/projects/personal/shopnest/node_modules/.pnpm/metro@0.76.7/node_modules/metro/src/DeltaBundler/Graph.js:473:35)
    at Graph._processModule (/home/sandheep/projects/personal/shopnest/node_modules/.pnpm/metro@0.76.7/node_modules/metro/src/DeltaBundler/Graph.js:261:38)
    at async Graph._traverseDependenciesForSingleFile (/home/sandheep/projects/personal/shopnest/node_modules/.pnpm/metro@0.76.7/node_modules/metro/src/DeltaBundler/Graph.js:249:5)
    at async Promise.all (index 0)
    at async Graph.initialTraverseDependencies (/home/sandheep/projects/personal/shopnest/node_modules/.pnpm/metro@0.76.7/node_modules/metro/src/DeltaBundler/Graph.js:233:5)
    at async DeltaCalculator._getChangedDependencies (/home/sandheep/projects/personal/shopnest/node_modules/.pnpm/metro@0.76.7/node_modules/metro/src/DeltaBundler/DeltaCalculator.js:229:25)
    at async DeltaCalculator.getDelta (/home/sandheep/projects/personal/shopnest/node_modules/.pnpm/metro@0.76.7/node_modules/metro/src/DeltaBundler/DeltaCalculator.js:112:16)
    at async DeltaBundler.buildGraph (/home/sandheep/projects/personal/shopnest/node_modules/.pnpm/metro@0.76.7/node_modules/metro/src/DeltaBundler.js:54:5)
    at async IncrementalBundler.buildGraphForEntries (/home/sandheep/projects/personal/shopnest/node_modules/.pnpm/metro@0.76.7/node_modules/metro/src/IncrementalBundler.js:66:19)
    at async IncrementalBundler.buildGraph (/home/sandheep/projects/personal/shopnest/node_modules/.pnpm/metro@0.76.7/node_modules/metro/src/IncrementalBundler.js:146:19)
› Stopped server
```

### Package Manager Version

pnpm --version 8.9.2

### Operating System

- [ ] macOS
- [X] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

I am using Ubuntu 23.10

**Comment by medv** [comment]
Doesn't work with pnpm at this stage, had to use npm unfortunately.

**Comment by mansdahlstrom1** [comment]
Any update on this one? @medv is pnpm really the issue here?

**Comment by medv** [comment]
I'm away from the pc today, but search nx discord for pnpm and you will find a solution that worked for me. Forgot to come back here to post it

**Comment by medv** [comment]
@mansdahlstrom1 I'm away from the pc today, but search nx discord for pnpm and you will find a thread I started that contains a solution. Might come back here to post it tomorrow.

**Comment by medv** [comment]
@mansdahlstrom1  add `node-linker=hoisted` to `.npmrc` at the root of your monorepo, remove your node modules and do a pnpm install. Additionally, make a copy of that `.npmrc` in each of your expo apps directories so that EAS build can use the same pnpm config. You would also want to pin the node and pnpm versions in each of your expo apps' `eas.json`, their docs explain how to do this. 

This works with expo@49 using the following in `metro.config.js`:

```
unstable_enableSymlinks: false,
unstable_enablePackageExports: false,
```

expo@50 uses a newer metro that has targeted some pnpm fixes, so this will either be unnecessary or will need to be revised and adjusted accordingly. If you find that this doesn't work for 50  but figure it out, drop a line here for the rest of us!

**Comment by mansdahlstrom1** [comment]
thanks for the insight @medv. I ended up reverting moving to pnpm and went back to npm for now! Also I'm using React Native CLI not Expo.

**Comment by medv** [comment]
@mansdahlstrom1 no worries, this issue was originally about expo but this _should_ apply to anyone using metro with pnpm. If you are, update everything and follow metro's changelog/issues over the last couple of months - there are relevant improvements there. Meanwhile, the above (`.npmrc` containing `node-linker=hoisted`) and the two experimental features set to `false` in metro config) should work for you if you are on an older version of metro. AFAIK they formalized `enableSymlinks` and automatically enabled it which could impact this, or remove the entire need for hoisting, but I could be imagining it.

Will be looking into this within the next month, so will try to remember to return here and update (unless somebody beats me to it ;))

**Comment by medv** [comment]
expo 50 with pnpm no longer requires hoisted dependencies, however still exhibits other issues.
Bun workspaces work out of the box so far with the following metro.config.ts (note I am executing this with bun, you will have to change this to .js and replace `import` with `require` and `export default` with `module.exports = config`

with both pnpm and bun, metro cannot resolve package.json `"main": "expo-router/entry"`, instead the old way of `"main": "index.js"`  that just contains `import 'expo-router/entry` works well. 
There are still these teething issues with expo+metro left that don't quite end at what I've covered, but it's certainly more than usable and may save you a lot of time and effort compared to a nx integrated monorepo. I've decided to simply use workspaces with a good tsconfig setup instead of nx for now and it is much more breezy. We will get back into nx soon, but not in an integrated monorepo style, because the downsides are too numerous.
